projects/plugins/super-cache/wp-cache.php

```diff
@@ -3080,35 +3080,85 @@ function wpsc_get_htaccess_info() {
 	$gziprules =  "<IfModule mod_mime.c>\n  <FilesMatch \"\\.html\\.gz\$\">\n    ForceType text/html\n    FileETag None\n  </FilesMatch>\n  AddEncoding gzip .gz\n  AddType text/html .gz\n</IfModule>\n";
 	$gziprules .= "<IfModule mod_deflate.c>\n  SetEnvIfNoCase Request_URI \.gz$ no-gzip\n</IfModule>\n";
 
-	$vary_header = $cache_control_header = '';
+	// Default headers.
+	$headers = array(
+		'Vary'          => 'Accept-Encoding, Cookie',
+		'Cache-Control' => 'max-age=3, must-revalidate',
+	);
 
-	if ( defined( 'WPSC_VARY_HEADER' ) ) {
-		if ( WPSC_VARY_HEADER != '' ) {
-			$vary_header = WPSC_VARY_HEADER;
-		}
-	} else {
-		$vary_header = 'Accept-Encoding, Cookie';
-	}
-	if ( defined( 'WPSC_CACHE_CONTROL_HEADER' ) ) {
-		if ( WPSC_CACHE_CONTROL_HEADER != '' ) {
-			$cache_control_header = WPSC_CACHE_CONTROL_HEADER;
-		}
-	} else {
-		$cache_control_header = 'max-age=3, must-revalidate';
+	// Allow users to override the Vary header with WPSC_VARY_HEADER.
+	if ( defined( 'WPSC_VARY_HEADER' ) && ! empty( WPSC_VARY_HEADER ) ) {
+		$headers['Vary'] = WPSC_VARY_HEADER;
 	}
-	$headers_text = "";
-	if ( $vary_header != '' ) {
-		$headers_text .= "  Header set Vary '$vary_header'\n";
-	}
-	if ( $cache_control_header != '' ) {
-		$headers_text .= "  Header set Cache-Control '$cache_control_header'\n";
+
+	// Allow users to override Cache-control header with WPSC_CACHE_CONTROL_HEADER
+	if ( defined( 'WPSC_CACHE_CONTROL_HEADER' ) && ! empty( WPSC_CACHE_CONTROL_HEADER ) ) {
+		$headers['Cache-Control'] = WPSC_CACHE_CONTROL_HEADER;
 	}
+
+	// Allow overriding headers with a filter.
+	$headers = apply_filters( 'wpsc_htaccess_mod_headers', $headers );
+
+	// Combine headers into a block of text.
+	$headers_text = join(
+		"\n",
+		array_map(
+			function ( $key, $value ) {
+				return "  Header set $key '" . addcslashes( $value, "'" ) . "'";
+			},
+			array_keys( $headers ),
+			array_values( $headers )
+		)
+	);
+
+	// Pack headers into gziprules (for historic reasons) - TODO refactor the values
+	// returned to better reflect the blocks being written.
 	if ( $headers_text != '' ) {
-		$gziprules .= "<IfModule mod_headers.c>\n$headers_text</IfModule>\n";
+		$gziprules .= "<IfModule mod_headers.c>\n$headers_text\n</IfModule>\n";
+	}
+
+	// Guess an appropriate ExpiresByType value based on the cache control header
+	$expires_by_type = 'A3';
+	if ( preg_match( '/max-age=(\d+)/', $headers['Cache-Control'], $matches ) ) {
+		$expires_by_type = 'A' . $matches[1];
```

Comment:
IMHO, since we are now providing filters, it may be better not to add this functionality of setting expiration headers based on the constant. That would discourage new adoption of the constant.

I also think this should be considered a corner case.

Reply:
@eAdnan007 , actually this really needs to make it into Super Cache. It will fix the issue with a possible misconfiguration [described here](https://wordpress.org/support/topic/allow-filtering-the-htaccess-entry-expiresbytype-text-html-for-expert-mode/#post-16273621).